module/zfs/dmu.c

```diff
@@ -2273,19 +2280,22 @@ dmu_brt_clone(objset_t *os, uint64_t object, uint64_t offset, uint64_t length,
 		ASSERT(db->db_blkid != DMU_BONUS_BLKID);
 		ASSERT(BP_IS_HOLE(bp) || dbuf->db_size == BP_GET_LSIZE(bp));
 
-		if (db->db_state == DB_UNCACHED) {
-			/*
-			 * XXX-PJD: If the dbuf is already cached, calling
-			 * dmu_buf_will_not_fill() will panic on assertion
-			 * (db->db_buf == NULL) in dbuf_clear_data(),
-			 * which is called from dbuf_noread() in DB_NOFILL
-			 * case. I'm not 100% sure this is the right thing
-			 * to do, but it seems to work.
-			 */
-			dmu_buf_will_not_fill(dbuf, tx);
+		mutex_enter(&db->db_mtx);
+		VERIFY(!dbuf_undirty(db, tx));
+		ASSERT(list_head(&db->db_dirty_records) == NULL);
+		mutex_exit(&db->db_mtx);
+
+		if (db->db_buf != NULL) {
+			arc_buf_destroy(db->db_buf, db);
+			db->db_buf = NULL;
 		}
```

Comment:
Did you intentionally move this out of db_mtx?  According to comment in dmu_impl.h db_buf is protected by db_mtx, and I see cases where arc_buf_destroy() is called while holding it, so it is allowed.